### ISSUE TITLE
Enhancement Request: Layer Blending - Canvas

### ISSUE THREAD

**Issue by kppankratz** [smeighan/xLights#1106]
I've been playing around with the new layer blending option using the off effect to bulk affect layers below it.  Very nice, thank you!

One enhancement I'd like you to consider adding to it is the ability for the user to specify the number of layers that are affected by the new layer blending.  Right now every layer is affected as you know.  With the ability to specify the layers affected, the user would be able to isolate a group say from a background.

I've attached a video showing the atom preset I shared set on top of a static background.  But when the atom is set to rotate using the off effect in transparent mode, the background does as well.  If there was the option to define the number of layers affected only the first 10 (off effect & atom layers) would be manipulated, leaving the 11th layer alone to produce a static background or anything below those layers.

Thank you,
Kevin

[Rotating Atom.mov.zip](https://github.com/smeighan/xLights/files/1827346/Rotating.Atom.mov.zip)

**Comment by AzGilrock** [comment]
Yeah that's somewhat close to some of the goals we already have.  My idea was to be able to create "scenes".  Like layers 5&6 combine to scene 1 and layers 2&3&4 combine to scene 2 and then layer 1 can blend the 2 scenes.

**Comment by kppankratz** [comment]
I had a chance to play around with the most recent enhancement to the Canvas Layer blending under Windows (waiting on Dan's return to update the Mac build).

When specific layers are selected I noticed that everything else is ignored above or below the grouped layers.  This leaves a blank (black) background instead of showing anything else that might be overlapping within that time frame.

The attached video shows how the background image disappears as the Canvas Layer blending is being performed.  Any way to allow other layers to show through?

[Canvas Layers.mp4.zip](https://github.com/smeighan/xLights/files/1921685/Canvas.Layers.mp4.zip)

**Comment by keithsw1111** [comment]
That was deliberate Kevin ... in fact I even think I added a comment to say so in the code.

In part it was laziness. I didnt know what the impact would be of trying to get the under layer data into the buffer at that time ... and testing it all to work it out seemed hard ... so i put it out of scope.

It may not be that hard to do ... but ... doing so could further slow down rendering as well. To be honest I am not sure that is true but it did seem a risk ... so again i decided to limit the blend to just layers in the same model.

I guess what I had in my mind is an assumption that most people doing this would likely be trying to create a single model effect and the cross model blending just didnt seem to be a priority.

**Comment by kppankratz** [comment]
I'm a little confused by the cross model blending you are referring to?  Are you talking about cross layer blending on the model?  That would make more sense.

The video I attached earlier is a single model with a Butterfly effect background.  When placing the atom preset and using the canvas blend, the background or anything else for that matter, is essentially deleted during that time frame on the model.  A little limiting if you always have to take into consideration the background will be black, but not a game ender.

**Comment by keithsw1111** [comment]
So here is the problem ... having blended everything using the off effect and the canvas layer you have a problem ... you cant let everything underneath re-merge into the canvas layer as they would not work. Consequently the way the canvas layer works is it renders all the included layers before the off effect is rendered ... which just passes the data through ... but it then gets rendered as a layer 1 blend ... hiding everything below it.

To achieve what you are after I think you need to put the butterfly on top, the off next and then all the other bits below that. The secret is then the blend style you set on the butterfly ... I think.

I am not sure there is a better way to do it.

For now I would consider this feature speculative and understand that we may agree to change it ... at which point existing effects may break and i am not going to implement automatic fixes. Just a note of caution.

**Comment by kppankratz** [comment]
After following your suggestions, I did get it to work better, along with a better understanding of how this is all working behind the scenes.  It's too bad that that multiple instances of the canvas blending cannot be used at the same time since it hides everything below it.

On modification, and I can open another ticket if needed...

Can X & Y position sliders with value curve options be added to the off effect?  Maybe only visible if transparent is selected AND canvas blend mode as well?  The attached video shows the canvas position changing, but I had to do that with VC's on the buffer.  I'm thinking the position sliders might be easier?  Just a thought.

Should I close this ticket now?
